sw/device/lib/dif/dif_spi_host_unittest.cc

```diff
@@ -0,0 +1,517 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_spi_host.h"
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/base/testing/mock_mmio.h"
+// TODO(#9027): Move GlobalMock to a more universal location.
+#include "sw/device/silicon_creator/testing/mask_rom_test.h"
+
+#include "spi_host_regs.h"  // Generated.
+
+namespace dif_spi_host_unittest {
+namespace {
+
+// Mock out the spi_host_fifo functions.
+namespace internal {
+class MockFifo : public ::mask_rom_test::GlobalMock<MockFifo> {
+ public:
+  MOCK_METHOD(void, write, (const dif_spi_host_t *, const void *, uint16_t));
+  MOCK_METHOD(void, read, (const dif_spi_host_t *, void *, uint16_t));
+};
+}  // namespace internal
+using MockFifo = testing::StrictMock<internal::MockFifo>;
+extern "C" {
+void spi_host_fifo_write_alias(const dif_spi_host_t *spi_host, const void *src,
+                               uint16_t len) {
+  MockFifo::Instance().write(spi_host, src, len);
+}
+void spi_host_fifo_read_alias(const dif_spi_host_t *spi_host, void *dst,
+                              uint16_t len) {
+  MockFifo::Instance().read(spi_host, dst, len);
+}
+}
+
+using mock_mmio::MmioTest;
+using mock_mmio::MockDevice;
+using testing::ElementsAre;
+using testing::Test;
+
+#define ARRAYSIZE(x) (sizeof(x) / sizeof(x[0]))
```

Comment:
Does `memory.h` not compile as C++? It definitely should...

Reply:
No, it does not.  The definitions of stardard-library functions in `memory.h` (memcpy and friends) upset the C++ compiler.